src/ccl_background.c

```diff
@@ -621,6 +621,150 @@ void ccl_cosmology_compute_distances(ccl_cosmology * cosmo, int *status)
   }
 }
 
+/* ----- ROUTINE: ccl_cosmology_distances_from_input ------
+INPUT: cosmology, scale factor array, comoving distance chi computed at the scale factor array values
+TASK: if not already there, make a table of comoving distances from an input array
+*/
+void ccl_cosmology_distances_from_input(ccl_cosmology * cosmo, int na, double a[], double chi_a[], double E_a[],
+   int *status)
+{
+  //Do nothing if everything is computed already
+  if(cosmo->computed_distances)
+    return;
+
+  // Allocate E(a), chi(a) and a(chi) splines
+  gsl_spline * E = gsl_spline_alloc(cosmo->spline_params.A_SPLINE_TYPE, na);
+  gsl_spline * chi = gsl_spline_alloc(cosmo->spline_params.A_SPLINE_TYPE, na);
+  gsl_spline * achi = gsl_spline_alloc(cosmo->spline_params.A_SPLINE_TYPE, na);
+
+  //Check for too little memory
+  if (a == NULL || E_a == NULL || chi_a == NULL || E == NULL || chi == NULL || achi == NULL) {
+    *status=CCL_ERROR_MEMORY;
+    ccl_cosmology_set_status_message(
+      cosmo, "ccl_background.c: ccl_cosmology_distances_from_input(): ran out of memory\n");
+  }
+  /*
+  //Check for messed up scale factor conditions
+  if (!*status){
+    if ((a[0] < cosmo->spline_params.A_SPLINE_MINLOG) ||
+        (a[na-1]-1.0 > 1e-8)) {
+      *status = CCL_ERROR_LINSPACE;
+      ccl_cosmology_set_status_message(
+        cosmo,
+        "ccl_background.c: ccl_cosmology_compute_distances(): Error "
+        "creating first logarithmic and then linear spacing in a\n");
+    }
+  }*/
+
+  // Initialize a E(a) spline
+  if (!*status){
+    if (gsl_spline_init(E, a, E_a, na)){
+      *status = CCL_ERROR_SPLINE;
+      ccl_cosmology_set_status_message(
+        cosmo, "ccl_background.c: ccl_cosmology_distances_from_input(): Error creating  E(a) spline\n");
+    }
+  }
+
+  // Initialize chi(a) spline
+  if (!*status){
+    if (gsl_spline_init(chi, a, chi_a, na)){//in Mpc
+      *status = CCL_ERROR_SPLINE;
+      ccl_cosmology_set_status_message(
+        cosmo, "ccl_background.c: ccl_cosmology_distances_from_input(): Error creating  chi(a) spline\n");
+    }
+  }
+
+  // Reverse the order of chi(a) so that we can initialize the a(chi) spline, which needs monotonically decreasing x-array.
+  double chi_a_reversed[na], a_reversed[na];
```

Comment:
This may get you into trouble if `na` is too large, so unfortunately we should allocate these dynamically.

Reply:
Also, this blocked should be wrapped inside an `if (!*status)` block.

Reply:
Would that be something like this? 

```suggestion
  chi_a_reversed = malloc(sizeof(double)*na);
  a_reversed = malloc(sizeof(double)*na);
```

Also, to update the A_SPLINE_MIN, you mean to update this value to the lowest value of the input scale factor by the user, (i.e. cosmo->spline_params.A_SPLINE_MIN = a[0])?

Reply:
Yes, and yes! You'll have to check that malloc has worked (i.e. check that the arrays are not NULL), and then rememer to free them up afterwards.